File: cpp/Mod_tokenize.cpp
#pragma once

List<Tuple2<String, String>> tokenize(String formel){
	List<Char> allowed = {'a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z','0','1','2','3','4','5','6','7','8','9'};
	List<Char> formChar = s2lc(formel);
	List<Tuple2<String, String>> res;
	
	for(int i=0; i<length(formChar); i++){
		String token;
		String in;
		switch(get(formChar, i)){
			case '(':
				token = "PO";
				in = "(";
				break;
			case ')':
				token = "PC";
				in = ")";
				break;
			case '0':
				token = "FALSE";
				in = "0";
				break;
			case '1':
				token = "TRUE";
				in = "1";
				break;
			case '!':
				token = "NOT";
				in = "!";
				break;
			case '&':
				token = "AND";
				in = "&&";
				i++;
				break;
			case '|':
				token = "OR";
				in = "||";
				i++;
				break;
			case '-':
				token = "IMPLY";
				in = "->";
				i++;
				break;
			case ' ':
				continue;
			default:
				token = "VAR";
				break;
		}
		
		if(token == "VAR"){
			while(contains(allowed, get(formChar, i))){
				in = append(in, get(formChar, i));
				i++;
			}
			i--;
		}
		res = append(res, tuple2(token, in));
	}
	return res;
}

File: hs/Mod_tokenize.hs
module Mod_tokenize ( tokenize ) where
import Prelude hiding (and, or)
import Debug.Trace
import Splib
import Types

allowed = ['0'..'9'] ++ ['a'..'z']

tokenize :: String -> [(String, String)]
tokenize formel = tokenize_loop formel 0 []

tokenize_loop :: String -> Int -> [(String, String)] -> [(String, String)]
tokenize_loop formel i res
 | i == length formel = res 
 | next == '(' = tokenize_loop formel (i+1) (res++[("PO", "(")])
 | next == ')' = tokenize_loop formel (i+1) (res++[("PC", ")")])
 | next == '0' = tokenize_loop formel (i+1) (res++[("FALSE", "0")])
 | next == '1' = tokenize_loop formel (i+1) (res++[("TRUE", "1")])
 | next == '!' = tokenize_loop formel (i+1) (res++[("NOT", "!")])
 | next == '&' = tokenize_loop formel (i+2) (res++[("AND", "&&")])
 | next == '|' = tokenize_loop formel (i+2) (res++[("OR", "||")])
 | next == '-' = tokenize_loop formel (i+2) (res++[("IMPLY", "->")])
 | next == ' ' = tokenize_loop formel (i+1) res
 | otherwise = (tokenize_loop formel (offset) (res++[("VAR", var)]))
  where 
    offset = get21 (parseVar formel i [])
    var = get22 (parseVar formel i [])
    next = get formel i

parseVar :: String -> Int -> String -> (Int, String)
parseVar formel i res 
 | contains allowed next = parseVar formel (i+1) (res++[next])
 | otherwise = ((i), res)
  where
   next = get formel i